src/Storages/ObjectStorageQueue/StorageObjectStorageQueue.cpp

```diff
@@ -175,18 +182,20 @@ StorageObjectStorageQueue::StorageObjectStorageQueue(
     LOG_INFO(log, "Using zookeeper path: {}", zk_path.string());
     task = getContext()->getSchedulePool().createTask("ObjectStorageQueueStreamingTask", [this] { threadFunc(); });
 
-    /// Get metadata manager from ObjectStorageQueueMetadataFactory,
-    /// it will increase the ref count for the metadata object.
-    /// The ref count is decreased when StorageObjectStorageQueue::drop() method is called.
-    files_metadata = ObjectStorageQueueMetadataFactory::instance().getOrCreate(zk_path, *queue_settings);
-    try
-    {
-        files_metadata->initialize(configuration_, storage_metadata);
-    }
-    catch (...)
+    ObjectStorageQueueTableMetadata table_metadata(
+        *queue_settings, storage_metadata.getColumns(),
+        configuration_->format, processing_threads_num_from_cpu_cores);
+
+    files_metadata = ObjectStorageQueueMetadataFactory::instance().getOrCreate(
+        zk_path,
+        std::make_shared<ObjectStorageQueueMetadata>(zk_path, std::move(table_metadata), queue_settings),
+        *queue_settings);
+
+    if (processing_threads_num_from_cpu_cores)
     {
-        ObjectStorageQueueMetadataFactory::instance().remove(zk_path);
-        throw;
+        engine_args->settings->as<ASTSetQuery>()->changes.insertSetting(
```

Comment:
Now we modify the setting in ast only after creating a metadata instance, because this setting could be changed there